src/workerd/api/worker-rpc.c++

```diff
@@ -2,19 +2,401 @@
+  inline kj::OneOf<v8::Local<v8::Function>, kj::String> tryGetFn(
+      Worker::Lock& lock,
+      kj::Maybe<kj::StringPtr> entrypointName,
+      IoContext& ctx,
+      kj::StringPtr methodName) {
+
+    auto& handler = KJ_REQUIRE_NONNULL(lock.getExportedHandler(entrypointName, ctx.getActor()),
+        "Failed to get handler to worker.");
+
+    auto handle = handler.self.getHandle(lock);
+    if (verifyCallable(lock, handle, methodName)) {
+      return kj::str("'", methodName, "' is a registered method and cannot be called over RPC."_kj);
+    }
+    auto methodStr = jsg::v8StrIntern(lock.getIsolate(), methodName);
+    auto fnHandle = jsg::check(handle->Get(lock.getContext(), methodStr));
+
+    // Get() doesn't handle private properties so we have to get this separately.
+    auto priv = v8::Private::New(lock.getIsolate(), methodStr);
+    auto privHandle = jsg::check(handle->GetPrivate(lock.getContext(), priv));
+
+    // TODO(someday): The interceptor on the client side doesn't support private property
+    // interception, so the check is essentially dead code today. We'll check anyways to guard
+    // against that changing in the future.
+    if (!fnHandle->IsFunction() || (!privHandle.IsEmpty() && privHandle->IsPrivate())) {
+      // We can't return this because it's not a function, or if it is, it's private.
+      return kj::str("The RPC receiver does not implement the requested method."_kj);
+    }
+    return fnHandle.As<v8::Function>();
+  }
+
+  // Deserializes the arguments and passes them to the given function.
+  v8::Local<v8::Value> invokeFn(
+      jsg::Lock& js,
+      v8::Local<v8::Function> fn,
+      kj::ArrayPtr<const kj::byte> serializedArgs) {
+    // We received arguments from the client, deserialize them back to JS.
+    if (serializedArgs.size() > 0) {
+      auto args = KJ_REQUIRE_NONNULL(
+          deserializeV8(js, kj::heapArray(serializedArgs)).tryCast<jsg::JsArray>(),
+          "expected JsArray when deserializing arguments.");
+      // Call() expects a `Local<Value> []`... so we populate an array.
+      auto arguments = kj::heapArray<v8::Local<v8::Value>>(args.size());
+      for (size_t i = 0; i < args.size(); ++i) {
+        arguments[i] = args.get(js, i);
+      }
+      return jsg::check(fn->Call(js.v8Context(), fn, args.size(), arguments.begin()));
+    } else {
+      return jsg::check(fn->Call(js.v8Context(), fn, 0, nullptr));
+    }
+  };
+
+  // The serialized JS value and a bool to discern whether we've thrown an exception.
+  // Serializing may throw if the data we are serializing is not structured-cloneable.
+  // In this case, we will just serialize the exception message and set `caughtException`.
+  struct SerializedResult {
+    kj::Array<kj::byte> data;
+    bool caughtException;
+  };
+
+  // Serializes the return value of our method.
+  SerializedResult serialize(jsg::Lock& js, jsg::JsValue value) {
+    kj::Maybe<SerializedResult> result;
+
+    js.tryCatch([&] {
```

Comment:
The default handling of an uncaught JS exception is already to stringify it in order to convert it into a KJ "tunneled" exception, which seems good enough here.

> Specifically referring to the stringification of the JS value that we failed to serialize, seems to be missing when I don't stringify the error message 🙁

I am not sure it's a good idea to stringify the unserializable value and try to include that in the error. That value could include PII, or it could be giantic, or quite likely it cannot be stringified anyway.

Reply:
https://github.com/cloudflare/workerd/compare/c3768d96f24fb10a9b3edda93995950d4e05e07d..381e9a9fafeb109efddbd902ceee42aa2aa95d54

I dropped `serialize()`, just using `serializeV8()` directly.

Note that this means if I try to return a value (from the remote DO) that failed to be serialized, the client catches `'Error: could not be cloned.'` from the server.

---

Since you mentioned tunneled exceptions + stringification, we can also do something like
- https://github.com/cloudflare/workerd/compare/381e9a9fafeb109efddbd902ceee42aa2aa95d54..a9254ed192e88344b38851274cb677fb36bd0fd6

which will stringify the failure to serialize error, but of course this runs into the concerns you mentioned (PII, size, failure to stringify).

Perhaps it's worth mentioning that if a client attempts to call a method over RPC and tries to pass an argument that cannot be serialized, it won't necessarily throw `'Error: could not be cloned.'`; it'll throw something like `'DataCloneError: <stringified representation of JS object> could not be cloned.'`. Maybe this isn't as much of a concern because its all local to the client anyways (not sending data over network + client isn't seeing something that might be private to the server).

@jasnell do you have an opinion here? If serializing the result on the server side fails, our two options are:
1.  we always throw `'Error: could not be cloned.'` from the server, or 
2. we attempt to throw the stringified JS object.

Option 1 and 2 are (respectively) implemented in the two diffs I included.